File: src/main/java/com/mateusz/spring_boot_todos/services/JpaUserDetailsService.java
package com.mateusz.spring_boot_todos.services;

import org.springframework.context.annotation.Lazy;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import com.mateusz.spring_boot_todos.models.SecurityUser;
import com.mateusz.spring_boot_todos.repositories.UserRepository;

@Service
public class JpaUserDetailsService implements UserDetailsService {

    private final UserRepository userRepository;

    public JpaUserDetailsService(@Lazy UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
        return userRepository
                .findByUsername(username)
                .map(SecurityUser::new)
                .orElseThrow(() -> new UsernameNotFoundException("Username not found: " + username));
    }

}

File: src/main/kotlin/com/mateusz/todos/kotlin_sboot_todos_backend/services/JpaUserDetailsService.kt
package com.mateusz.todos.kotlin_sboot_todos_backend.services

import com.mateusz.todos.kotlin_sboot_todos_backend.models.SecurityUser
import com.mateusz.todos.kotlin_sboot_todos_backend.repositories.UserRepository
import org.springframework.context.annotation.Lazy
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class JpaUserDetailsService(@param:Lazy private val userRepository: UserRepository) :
    UserDetailsService {
    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(username: String): UserDetails {
        return userRepository
            .findByUsername(username)
            ?.map { SecurityUser(it!!) }
            ?.orElseThrow { UsernameNotFoundException("Username not found: $username") } ?: throw UsernameNotFoundException("Username not found: $username")
    }
}
